### PR TITLE
Right click create group

### PR DESCRIPTION
**Pull request by m-peeler**
Added functionality to allow users to right click to make a new group from the selection. Updated all group creation and group editing to use the CreateGroupAction class, extending SimpleCommand so it would work with the right-click menu. Closes https://github.com/JabRef/jabref/issues/11449 and should be ready for merging in.

Getting this functionality extracted into a SimpleCommand (needed by the right-click menu) that could be called with the available states stored in StateManager required some pretty large changes and bug fixes related to synchronizing the various lists of active and selected entries.

This has the changes from #11453 included in it, and slightly improved due to a few issues I faced when doing the other edits.

Additionally, fixed a bug where when new entries were dropped into a list, the displayed number of entries did not update.

I'll update the documentation when this gets approved for merge.

image

Future improvements may include letting the user pick if they want a group created above / beside / below the current group, or at root. 

### Mandatory checks

- [X] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [X] Tests created for changes (if applicable)
- [X] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [X] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,7 +11,10 @@ Note that this project **does not** adhere to [Semantic Versioning](https://semv
 
 ### Added
 
+- We added the ability to add the current selection to a newly created group. [#11449](https://github.com/JabRef/jabref/issues/11449)
 - We made new groups automatically to focus upon creation. [#11449](https://github.com/JabRef/jabref/issues/11449)
+- We added the ability to right-click on selected entries and create a group using them. [#11449](https://github.com/JabRef/jabref/issues/11449)
```
Comment (koppor) [inline CHANGELOG.md:16]: I think, you based your new branch on your branch before - and not on main. This is OK, we can manage.

Pleae remove this line, because that is already in main (rephrased).

```suggestion
```

Strange that GitHubs shows the cahnge in `StateManager`. Main should include it.

File: src/main/java/module-info.java [hunk 2]
```diff
@@ -150,4 +150,5 @@
     requires dd.plist;
     requires mslinks;
     requires org.apache.httpcomponents.core5.httpcore5;
+    requires commons.beanutils;
```
Comment (koppor) [inline src/main/java/module-info.java:153]: Why was this added?

File: CHANGELOG.md (hunk 1 above)
Comment (koppor) [reply CHANGELOG.md:16]: Maybe merge `upstream/main`?

**Review by koppor — COMMENTED** [review]
Just a quick comment - it's late here.

Please check the output of the build pipeline (checkstyle, openrewrite)

---

Need to discuss with @JabRef/developers whether this context menu or the choice in the group creation dialog was better.

I didn't find the PR of the group creation dialog, otherwise I would have linked it.

**Comment by m-peeler** [comment]
Working to address the failed tests now!

File: src/main/java/module-info.java (hunk 2 above)
Comment (m-peeler) [reply src/main/java/module-info.java:153]: I am not sure - I think it was auto-imported by IntelliJ, which is how I also ended up with so many .* imports. I couldn't find any other mentions of an import from it, so I can remove it and see how the build goes.

File: CHANGELOG.md (hunk 1 above)
Comment (m-peeler) [reply CHANGELOG.md:16]: Will do - I kept trying to do that but the desktop client was having some issues with it.

File: src/main/java/module-info.java (hunk 2 above)
Comment (Siedlerchr) [reply src/main/java/module-info.java:153]: This looks like an observable import from the wrong package

**Review by koppor — CHANGES_REQUESTED** [review]
Discussion result: We have special context menus at the group pane. Also "Add selected entries to group". To have **consistency** here, we propose:

Left side: "Add subgroup using selected entries".  Below "Add subgroup"

grafik

**Comment by m-peeler** [comment]
> Discussion result: We have special context menus at the group pane. Also "Add selected entries to group". To have **consistency** here, we propose:
> 
> Left side: "Add subgroup using selected entries". Below "Add subgroup"

That sounds good and should be fairly easily implemented with the changes I've made. I imagine it should be greyed out if there are multiple groups selected?

**Comment by m-peeler** [comment]
This push should close out the changes, add add in @koppor's request for the addition to the second context menu. Let me know if you find any issues.

File: src/main/java/org/jabref/gui/actions/ActionHelper.java [hunk 3]
```diff
@@ -43,6 +46,28 @@ public static BooleanExpression needsStudyDatabase(StateManager stateManager) {
         return BooleanExpression.booleanExpression(binding);
     }
 
+    // Makes sure there is at least one group selected, and if there are multiple groups selected
+    // all have the same parent node.
+    public static BooleanExpression needsSelectedGroupsShareParent(StateManager stateManager) {
+        if (stateManager.activeGroupProperty().isEmpty()) {
```
Comment (koppor) [inline src/main/java/org/jabref/gui/actions/ActionHelper.java:52]: This reads strange. A `BooleanExpression` is someething evaluated "live" during the run of the program. With the currente change, this live property is different depending on a condition on its creation.

Maybe, the whole `if` can be done. The `return` thing reads OK.

File: src/main/java/org/jabref/gui/actions/ActionHelper.java [hunk 4]
```diff
@@ -43,6 +46,28 @@ public static BooleanExpression needsStudyDatabase(StateManager stateManager) {
...
+            return Bindings.size(stateManager.activeGroupProperty()).isEqualTo(1);
+        }
+        return Bindings.size(stateManager.activeGroupProperty()).isEqualTo(1)
+           .or(Bindings.size(stateManager.activeGroupProperty()).greaterThan(1).and(
+                Bindings.size(
+                        (ObservableList<GroupTreeNode>) stateManager
+                                .activeGroupProperty()
+                                .stream()
+                                .filter(stateManager.activeGroupProperty().get().getFirst()::equals)
+                                .collect(Collectors.toCollection(FXCollections::observableArrayList))
```
Comment (koppor) [inline src/main/java/org/jabref/gui/actions/ActionHelper.java:62]: Reads strange. This is an active query. Can't one have a "passive" list build dynamically?

The "dynamics" of bindings are explained well here: https://docs.oracle.com/javafx/2/binding/jfxpub-binding.htm

File: src/main/java/org/jabref/gui/StateManager.java [hunk 5]
```diff
@@ -75,7 +75,13 @@ public class StateManager {
     private final ObservableList<String> searchHistory = FXCollections.observableArrayList();
 
     public StateManager() {
-        activeGroups.bind(Bindings.valueAt(selectedGroups, activeDatabase.orElseOpt(null).map(BibDatabaseContext::getUid)));
```
Comment (koppor) [inline src/main/java/org/jabref/gui/StateManager.java:78]: Hasn't JavaFX the option to have "Opt..."? The `NULL_CONTEXT` is a nice idea, but we never use that in JabRef. We have Optionals.

File: src/main/java/org/jabref/gui/maintable/RightClickMenu.java [hunk 6]
```diff
@@ -74,6 +74,14 @@ public static ContextMenu create(BibEntryTableViewModel entry,
                 SpecialFieldMenuItemFactory.getSpecialFieldSingleItem(SpecialField.PRINTED, factory, () -> libraryTab, dialogService, preferencesService, undoManager, stateManager),
                 SpecialFieldMenuItemFactory.createSpecialFieldMenu(SpecialField.PRIORITY, factory, () -> libraryTab, dialogService, preferencesService, undoManager, stateManager),
                 SpecialFieldMenuItemFactory.createSpecialFieldMenu(SpecialField.READ_STATUS, factory, () -> libraryTab, dialogService, preferencesService, undoManager, stateManager),
+                factory.createMenuItem(
+                        // The new group is placed inside the currently active group, so
+                        // the button's name changes based on whether we are in the root group
+                        // or a subgroup.
```
Comment (koppor) [inline src/main/java/org/jabref/gui/maintable/RightClickMenu.java:80]: Althrough I like this, it is consistent to    The "All entries" group always has "Add subgroup". See screenshot https://github.com/JabRef/jabref/pull/11476#pullrequestreview-2170254425

One would have to change the whole menu to be consistent. This is too much effort. Thus, alwys add subgroup...

**Review by koppor — CHANGES_REQUESTED** [review]
Quick first comments. Maybe, it fosters discusions / actions.

File: src/main/java/org/jabref/gui/actions/ActionHelper.java (hunk 3 above)
Comment (m-peeler) [reply src/main/java/org/jabref/gui/actions/ActionHelper.java:52]: That was an (admittedly short-sighted) attempt to get around the fact that the `.and()` function does not short-circuiting, so even if it failed the check for `size(activeGroupProperty()).greaterThan(1)` it would try to get the first element of the list and raise a null-pointer exception. As best I can tell with my debugger anyways, both the context menus in the Groups side view and the Main Table are re-created each time a right-click is registered, so I went ahead and got around the issue with the non-dynamic `if`. Looking closer, it does seem that the original `Add Groups` button at the bottom is the only one that is static in it's implementation (though since it's set to always add to root, that technically wouldn't be an issue).

File: src/main/java/org/jabref/gui/StateManager.java (hunk 5 above)
Comment (m-peeler) [reply src/main/java/org/jabref/gui/StateManager.java:78]: I was running into issues when returning `activeGroups.readOnlyProperty()` where it was returning out a `null` and causing problems. Upon review, I think I can achieve the same effect with the following:
```
    public StateManager() {
        activeGroups.bind(Bindings.valueAt(
            selectedGroups,
            activeDatabase
                .orElseOpt(null)
                .map(BibDatabaseContext::getUid))
            .orElse(FXCollections.observableArrayList()));
    }
```

File: src/main/java/org/jabref/gui/actions/ActionHelper.java (hunk 4 above)
Comment (m-peeler) [reply src/main/java/org/jabref/gui/actions/ActionHelper.java:62]: I believe this should fix the problem, and get rid of the if-statement.
```
    public static BooleanExpression needsSelectedGroupsShareParent(StateManager stateManager) {
        return Bindings.size(stateManager.activeGroupProperty()).isEqualTo(1)        // Is single element OR
            .or(Bindings.size(stateManager.activeGroupProperty()).greaterThan(1)     // (Is multiple elements AND
                .and(Bindings.createBooleanBinding(
                    () -> stateManager.activeGroupProperty().stream()
                        .allMatch(val ->                                                // The parent of the first element
                            stateManager.activeGroupProperty()                          // is equal to the parent of all
                                .getFirst()                                             // other elements 
                                .getParent()                                            // (including if they are null))
                                    .map(node -> node.equals(val.getParent().get()))
                                    .orElseGet(() -> val.getParent().isEmpty())),
                stateManager.activeGroupProperty())));
    }
```

File: src/main/java/org/jabref/gui/actions/ActionHelper.java [hunk 7]
```diff
@@ -43,6 +43,26 @@ public static BooleanExpression needsStudyDatabase(StateManager stateManager) {
         return BooleanExpression.booleanExpression(binding);
     }
 
+    // Makes sure there is at least one group selected, and if there are multiple groups selected
+    // all have the same parent node.
```
Comment (koppor) [inline src/main/java/org/jabref/gui/actions/ActionHelper.java:47]: Convert to JavaDoc

File: src/main/java/org/jabref/gui/actions/StandardActions.java [hunk 8]
```diff
@@ -36,6 +36,8 @@ public enum StandardActions implements Action {
     EXTRACT_FILE_REFERENCES_OFFLINE(Localization.lang("Extract references from file (offline)"), IconTheme.JabRefIcons.FILE_STAR),
     OPEN_URL(Localization.lang("Open URL or DOI"), IconTheme.JabRefIcons.WWW, KeyBinding.OPEN_URL_OR_DOI),
     SEARCH_SHORTSCIENCE(Localization.lang("Search ShortScience")),
+    ADD_GROUP_FROM_SELECTION(Localization.lang("Add group from selection")),
```
Comment (koppor) [inline src/main/java/org/jabref/gui/actions/StandardActions.java:39]: This can be removed? -- (I don't find the link to the comment I gave, maybe you remember the screenshot that "All entries" always have the context menu with "subgroups", not" groups")

File: src/main/java/org/jabref/gui/exporter/ExportCommand.java [hunk 9]
```diff
@@ -111,7 +111,7 @@ private void export(Path file, FileChooser.ExtensionFilter selectedExtensionFilt
             // All entries
             entries = stateManager.getActiveDatabase()
                                   .map(BibDatabaseContext::getEntries)
-                                  .orElse(Collections.emptyList());
```
Comment (koppor) [inline src/main/java/org/jabref/gui/exporter/ExportCommand.java:114]: Is this change needed? I checked the caller - and it seems it is passed to `org.jabref.logic.exporter.Exporter#export(org.jabref.model.database.BibDatabaseContext, java.nio.file.Path, java.util.List<org.jabref.model.entry.BibEntry>, java.util.List<java.nio.file.Path>, org.jabref.logic.journals.JournalAbbreviationRepository)`, which requires a list and nothing observable.

I would even try to use `toList()` since the list should not be modified (should it?)

**Review by koppor — CHANGES_REQUESTED** [review]
General comments before I comment on details in the code. I think, following comments will lead to much code removal and code change.

image

This one should be selected:

image

When the user presses "OK", the entries in the main table are then already in the group.

---

Remove the context menu entry in the main table

image

Reason: If it was kept, one also would need two other entries in the context menu:

image

But then, the menu is too large. We removed it at https://github.com/JabRef/jabref/pull/4695 - and no users complained since 2019. Thus, no change there.

File: src/main/java/org/jabref/gui/groups/GroupTreeView.java [hunk 10]
```diff
@@ -570,7 +599,7 @@ private ContextMenu createContextMenuForGroup(GroupNodeViewModel group) {
     }
 
     private void addNewGroup() {
-        viewModel.addNewGroupToRoot();
```
Comment (koppor) [inline src/main/java/org/jabref/gui/groups/GroupTreeView.java:573]: Please think of two options:

a) ENUM covering the options
b) two method (this is favored at [Java by Comparison](https://java.by-comparison.com/), item "Split Method with Boolean Parameters")

File: src/main/java/org/jabref/gui/exporter/ExportCommand.java (hunk 9 above)
Comment (m-peeler) [reply src/main/java/org/jabref/gui/exporter/ExportCommand.java:114]: I believe I just did this because the function signature changed (from `Optional<List<..>>` to `Optional<ObservableList<...>>` and I was trying to minimally fix to comply. Will add in the `toList` call and change it back.

**Comment by m-peeler** [comment]
> This one should be selected:
> 
> image
Currently, since there is usually one entry selected at all times, unless you ctrl-click it to unselect, I only have it adding the entries initially if they have multiple - that was if someone is trying to create an empty group it doesn't initially include entries they don't want. Should we maintain any sort of 'don't include selection' at any point, or just assume as such? If the latter is the case, it should lead to the removal of that boolean argument you wanted taken out.

**Comment by koppor** [comment]
> Currently, since there is usually one entry selected at all times, unless you ctrl-click it to unselect, I only have it adding the entries initially if they have multiple - that was if someone is trying to create an empty group it doesn't initially include entries they don't want. Should we maintain any sort of 'don't include selection' at any point, or just assume as such? If the latter is the case, it should lead to the removal of that boolean argument you wanted taken out.


I understood  your proposal as follows:

- User selectes "Add subgroup" -> group dialog opens, user work as sual
- User selects "Add subgroup from selected entries" -> group dialog opens, "explicit selection" selected. ~~User can use group dialog. User can change radio button. If may choose other group (e.g., keywords). Then, notification: "selected entries"..~~ User cannot choose other radio button. Meaning: user cannot change away from explicit selection. User enters data. User presses OK. Then, group with explicitly selection type is created. Selected entries added.

Reason:

- It is not generally possible to deduct the keywords or other queries for the selected entries in a "nice" way, so that the user is satisfied in 80% of the cases. Thus, group type "explicit selection"
- We trigger the explicit selection "externally" from the group dialog. My initial proposal was the checkbox inside the group dialog, which was not done. The implementation moved to some context menu in the main table, which we moved to the group panel.

Did I get something wrong?

**Comment by m-peeler** [comment]
> * We trigger the explicit selection "externally" from the group dialog. My initial proposal was the checkbox inside the group dialog, which was not done. The implementation moved to some context menu in the main table, which we moved to the group panel.

I think I somehow misread the original checkbox request as a radio button. Check box should be easy to add - I'll do that quickly - and should resolve the issues.

**Comment by m-peeler** [comment]
Conversion from radio button to checkbox (including the ability to set the preference) has been done.

File: src/main/java/org/jabref/gui/groups/GroupDialogViewModel.java [hunk 11]
```diff
@@ -78,6 +80,8 @@ public class GroupDialogViewModel {
     private final BooleanProperty typeTexProperty = new SimpleBooleanProperty();
 
     // Option Groups
```
Comment (koppor) [inline src/main/java/org/jabref/gui/groups/GroupDialogViewModel.java:82]: The heading does not match the following things.

Please move the Boolean Property.

File: src/main/java/org/jabref/gui/groups/GroupDialogViewModel.java [hunk 12]
```diff
@@ -112,24 +116,42 @@ public class GroupDialogViewModel {
...
 
+    public GroupDialogViewModel(
+            DialogService dialogService,
+            BibDatabaseContext currentDatabase,
+            PreferencesService preferencesService,
+            @Nullable AbstractGroup editedGroup,
+            @Nullable GroupTreeNode parentNode,
+            FileUpdateMonitor fileUpdateMonitor
+    ) {
+        this(dialogService, currentDatabase, preferencesService, editedGroup, parentNode, fileUpdateMonitor, new ArrayList<>(), Selection.IGNORE_SELECTED_ENTRIES);
```
Comment (koppor) [inline src/main/java/org/jabref/gui/groups/GroupDialogViewModel.java:152]: USe `List.of()` instead of `new ArrayList<>()`

File: src/main/java/org/jabref/gui/groups/GroupDialogViewModel.java [hunk 13]
```diff
@@ -310,10 +332,15 @@ public AbstractGroup resultConverter(ButtonType button) {
         try {
             String groupName = nameProperty.getValue().trim();
             if (typeExplicitProperty.getValue()) {
-                resultingGroup = new ExplicitGroup(
+                ExplicitGroup tempResultingGroup = new ExplicitGroup(
```
Comment (koppor) [inline src/main/java/org/jabref/gui/groups/GroupDialogViewModel.java:335]: Why a temp group?

You can directly add to a group?

Please change back and just add to `resultingGroup`

File: src/main/java/org/jabref/gui/groups/GroupNodeViewModel.java [hunk 14]
```diff
@@ -65,47 +65,60 @@ public class GroupNodeViewModel {
...
+            children.forEach(c -> groupNode.addChild(c.getGroupNode()));
         } else {
             children = EasyBind.mapBacked(groupNode.getChildren(), this::toViewModel);
         }
         if (groupNode.getGroup() instanceof TexGroup) {
             databaseContext.getMetaData().groupsBinding().addListener(new WeakInvalidationListener(onInvalidatedGroup));
         }
         hasChildren = new SimpleBooleanProperty();
-        hasChildren.bind(Bindings.isNotEmpty(children));
+        hasChildren.bind(Bindings.isNotEmpty(groupNode.getChildren()));
```
Comment (koppor) [inline src/main/java/org/jabref/gui/groups/GroupNodeViewModel.java:98]: The old code did not work?

**Review by koppor — CHANGES_REQUESTED** [review]
Quick comments. Need to be more awake for the rest.

File: src/main/java/org/jabref/gui/groups/GroupDialog.fxml [hunk 15]
```diff
@@ -102,6 +102,9 @@
                 </VBox>
                 <Separator orientation="VERTICAL"/>
                 <StackPane HBox.hgrow="ALWAYS">
+                    <VBox visible="${explicitRadioButton.selected}" spacing="10.0">
+                        <CheckBox fx:id="explicitIncludeSelected" text="%Include selected entries in new subgroup"/>
+                    </VBox>
```
[the remark this answers is not in the document]
Comment (m-peeler) [reply src/main/java/org/jabref/gui/groups/GroupDialog.fxml:107]: Here are the two changed menues
image
image

File: src/main/java/org/jabref/gui/groups/GroupDialogViewModel.java (hunk 13 above)
Comment (m-peeler) [reply src/main/java/org/jabref/gui/groups/GroupDialogViewModel.java:335]: `add` is only on `EntriesChanger`, which `ExplicitGroup` implements but `AbstractGroup` does not. I used the `temp` rather than casting for adding on line 341, but can cast if you prefer.

File: src/main/java/org/jabref/gui/groups/GroupNodeViewModel.java (hunk 14 above)
Comment (m-peeler) [reply src/main/java/org/jabref/gui/groups/GroupNodeViewModel.java:98]: Think there was a potential problem with the main table context menu that this was for, but since that's been removed I will revert.

**Comment by m-peeler** [comment]
(In reference to this comment - )[https://github.com/JabRef/jabref/pull/11476#discussion_r1676536347] this push changes the method signature (since it was already an ObservableList and something needed to be able to bind to it). Removing the `.map(List::stream).map(Stream::toList)` causes compile errors and seems to be the reason tests are currently failing. I've changed it to instead do `.orElse(FXCollections.emptyObservableList())` and that seems to address the compile issue.

**Review by koppor — CHANGES_REQUESTED** [review]
We have now both the possibility in the dialog to change the behavior and using the contextd menu

image

This is too much for this small functionality. We also store the default in the preferences (which is good)

Thus, please remove the context menu entry.

File: src/main/java/org/jabref/gui/exporter/ExportCommand.java (hunk 9 above)
Comment (m-peeler) [reply src/main/java/org/jabref/gui/exporter/ExportCommand.java:114]: Per my earlier comment - 

> This push changes the method signature of `getEntries()` (since it was already an `ObservableList` and something needed to be able to bind to it). Removing the `.map(List::stream).map(Stream::toList)` causes compile errors and seems to be the reason tests are currently failing. I've changed it to instead do `.orElse(FXCollections.emptyObservableList())` and that seems to address the compile issue.

**Comment by m-peeler** [comment]
> Thus, please remove the context menu entry.

Done. I also set the button to be disabled when editing a group rather than creating one.

**Comment by Siedlerchr** [comment]
@LoayGhreeb  Will this create conflicts with your search float mode PR?

**Comment by LoayGhreeb** [comment]
> Will this create conflicts with your search float mode PR?

Yeah, would be better to merge the floating mode PR first.

**Comment by koppor** [comment]
@m-peeler We took long to get https://github.com/JabRef/jabref/pull/11510 finished, but we finally managed.💪. Can you have a look whetehr you can resolve the merge conflicts so that we can get this PR finished, too?

**Comment by subhramit** [comment]
Ah, this will require conflict resolution again after Postgres search, I think?

**Comment by koppor** [comment]
> Ah, this will require conflict resolution again after Postgres search, I think?

Did not forsee that, but maybe yes. I see many conflicts.

**Comment by koppor** [comment]
Closing this issue due to inactivity :zzz: 

Please ping us if you intend to resume work on this one.
